### PR TITLE
python3Packages.torch: add descriptive messages when marked broken

### PR DESCRIPTION
**Pull request by ConnorBaker**
## Description of changes

Continuing some of the work started near the end of https://github.com/NixOS/nixpkgs/pull/249259, this PR adds a `brokenConditions` attribute to `passthru` to help provide insight into why `broken` in `meta` is true, without producing side-effects like emitting warnings.

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by SomeoneSerge** [comment]
I asked in the dev chat again and it does seem [preferable](https://matrix.to/#/!kjdutkOsheZdjqYmqp:nixos.org/$PmG7eyqSq8wuo4FDoNYW3wq-9_ScXZBbpm-CxIWaYE8?via=nixos.org&via=matrix.org&via=nixos.dev) that we keep `meta.broken` side effect-free.

What do you think about moving `conditions` to the `passthru`, so that when we encounter the failure we could still figure out the exact cause using `nix eval`?

**Comment by ConnorBaker** [comment]
I like that! Here's an example of how the latest works:

```console
$ nix build --impure .#python3Packages.torch
warning: Git tree '/home/connorbaker/nixpkgs' is dirty
error:
       … while evaluating the attribute 'drvPath'

         at /nix/store/g9wsfc0c72xfylz9mx6knjm4ji3n72nq-source/lib/customisation.nix:222:7:

          221|     in commonAttrs // {
          222|       drvPath = assert condition; drv.drvPath;
             |       ^
          223|       outPath = assert condition; drv.outPath;

       … while evaluating the attribute 'drvPath'

         at /nix/store/g9wsfc0c72xfylz9mx6knjm4ji3n72nq-source/lib/customisation.nix:222:7:

          221|     in commonAttrs // {
          222|       drvPath = assert condition; drv.drvPath;
             |       ^
          223|       outPath = assert condition; drv.outPath;

       (stack trace truncated; use '--show-trace' to show the full trace)

       error: Package ‘python3.10-torch-2.0.1’ in /nix/store/g9wsfc0c72xfylz9mx6knjm4ji3n72nq-source/pkgs/development/python-modules/torch/default.nix:449 is marked as broken, refusing to evaluate.

       a) To temporarily allow broken packages, you can use an environment variable
          for a single invocation of the nix tools.

            $ export NIXPKGS_ALLOW_BROKEN=1

        Note: For `nix shell`, `nix build`, `nix develop` or any other Nix 2.4+
        (Flake) command, `--impure` must be passed in order to read this
        environment variable.

       b) For `nixos-rebuild` you can set
         { nixpkgs.config.allowBroken = true; }
       in configuration.nix to override this.

       c) For `nix-env`, `nix-build`, `nix-shell` or any other Nix command you can add
         { allowBroken = true; }
       to ~/.config/nixpkgs/config.nix.
```

What a surprise! What could have gone wrong?

```console
$ nix eval --impure .#python3Packages.torch.brokenConditions
warning: Git tree '/home/connorbaker/nixpkgs' is dirty
{ "CUDA and ROCm are not mutually exclusive" = false; "CUDA is not targeting Linux" = false; "MPI cudatoolkit does not match cudaPackages.cudatoolkit" = false; "Magma cudaPackages does not match cudaPackages" = true; "Unsupported CUDA version" = false; }
```

Oh ;)

```nix
magma-cuda-static = magma-cuda.override {
  static = true;
  cudaPackages = cudaPackages_12;
};
```

File: pkgs/development/python-modules/torch/default.nix [hunk 1]
```diff
@@ -105,6 +105,14 @@ let
       rocm-runtime rocm-opencl-runtime hipify
     ];
   };
+
+  brokenConditions = {
```
Comment (SomeoneSerge) [inline pkgs/development/python-modules/torch/default.nix:109]: Might as well `filterAttrs (_: cond: cond)` to skip the passing tests from nix eval

**Review by SomeoneSerge — APPROVED** [review]
Discoverability suffers as is common, but 👍🏻 for internal use

File: pkgs/development/python-modules/torch/default.nix (hunk 1 above)
Comment (ConnorBaker) [reply pkgs/development/python-modules/torch/default.nix]: Good call; fixed in latest.

**Review by SomeoneSerge — APPROVED** [review]

**Pull request merged**